src/Features/Core/Portable/Completion/PatternMatchRule.cs

```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    internal class PatternMatchRule
+    {
+        public PatternMatchRuleKind Kind { get; }
+
+        public PatternMatchRule(PatternMatchRuleKind kind)
+        {
+            this.Kind = kind;
+        }
+    }
+
+    internal enum PatternMatchRuleKind
+    {
+        /// <summary>
+        /// Pick based on strength of a range of the item exactly matching the pattern 
+        /// </summary>
+        Exactness,
+
+        /// <summary>
+        /// Pick based on strength of the camel casing of the item matching the pattern
+        /// </summary>
+        CamelCasing, 
+
+        /// <summary>
+        /// Pick based on strength of the item matching the patterns casing
+        /// </summary>
+        CaseSensitivity,
+
+        /// <summary>
+        /// Pick based on strength of punctuation in the item matching the pattern
+        /// </summary>
+        Punctuation,
+
+        /// <summary>
+        /// Pick based on the item being an argument name
+        /// </summary>
+        ArgumentName,
```

Comment:
I would prefer if the last two are not in this list.  They shoudl be done within the service itself, not at the controller level.